core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4

```diff
@@ -284,7 +288,7 @@ joinCriteria
     ;
 
 sampledRelation
-    : aliasedRelation (
+    : (aliasedRelation | patternRecognitionRelation) (
```

Comment:
This introduces an ambiguity, since aliasedRelation is also a prefix of patternRecognitionRelation. The parser will likely need to backtrack if it decides to go down the aliasedRelation route and it encounters a MATCH_RECOGNIZE token after it. Antlr may be able to predict with a DFA, but it will require unbounded lookahead to make that decision, which makes the parsing less efficient.

You can see that in the grammar profiler in IntelliJ for a query such as:

```
(WITH N AS (SELECT * FROM A JOIN B ON 1 = 2) SELECT * FROM N ORDER BY N.F1) AS T MATCH_RECOGNIZE (PATTERN (()) 
```

you'll notice how the max lookahead grows as the query in the relationPrimary grows in complexity.

Reply:
I returned to the previous version of grammar here. It also has the benefit of showing `MATCH_RECOGNIZE` as an option in parse error.